k2/csrc/context.cuh

```diff
@@ -129,6 +131,86 @@ class Context : public std::enable_shared_from_this<Context> {
   virtual void Sync() const {}
 };
 
+// TODO(haowen): most of implementations below should be updated later.
+class CpuContext : public Context {
+ public:
+  ContextPtr GetCpuContext() override { return nullptr; }
+  ContextPtr GetPinnedContext() override { return nullptr; }
+  DeviceType GetDeviceType() const override { return kCpu; }
+
+  void *Allocate(std::size_t bytes, void **deleter_context) override {
+    void *p = nullptr;
+    if (bytes) {
+      int32_t ret = posix_memalign(&p, kAlignment, bytes);
+      K2_DCHECK_EQ(ret, 0);
+    }
+    if (deleter_context) *deleter_context = nullptr;
+    return p;
+  }
+
+  bool IsCompatible(const Context &other) const override {
+    return other.GetDeviceType() == kCpu;
+  }
+
+  void Deallocate(void *data, void * /*deleter_context*/) override {
+    free(data);
+  }
+};
+
+#ifdef __CUDACC__
+class CudaContext : public Context {
+ public:
+  CudaContext(int32_t gpu_id) : gpu_id_(gpu_id) {
+    if (gpu_id_ != -1) {
+      cudaSetDevice(gpu_id_);
```

Comment:
Are there any differences between the following two approches?

(1)
```cpp
auto ret = cudaSetDevice(gpu_id_);
K2_CHECK_CUDA_ERROR(ret);
```

(2)
```cpp
cudaSetDevice(gpu_id_);
K2_CHECK_CUDA_ERROR(cudaGetLastError());
```

----

I prefer the first one since it avoids one extra function call.

Reply:
I just thought it may be not just the performace issue (i.e. calling twice). If we call `cudaGetLastError` twice here (case 2, since the current implementation of `K2_CHECK` calls function twice) in a multiple thread case, we may get different error number. So we have to use the first one (case 1), or if we can avoid calling function twice in `K2_CHECK` by defining some intermediate variable, then we just call `K2_CUDA_CHECK_ERROR(cudaSetDevice(gpu_id))`  .